crypto/ocsp/ocsp_verify.c

```diff
@@ -0,0 +1,382 @@
+// with the same algorithm then there's no point trying to match any certificates
+// against the issuer, and |*ret| will be set to NULL.
+static int ocsp_check_ids(STACK_OF(OCSP_SINGLERESP) *sresp, OCSP_CERTID **ret) {
+  if (sresp == NULL || ret == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  OCSP_CERTID *tmpid, *cid;
+  size_t idcount = sk_OCSP_SINGLERESP_num(sresp);
+  if (idcount <= 0) {
+    OPENSSL_PUT_ERROR(OCSP, OCSP_R_RESPONSE_CONTAINS_NO_REVOCATION_DATA);
+    return -1;
+  }
+  cid = sk_OCSP_SINGLERESP_value(sresp, 0)->certId;
+
+  *ret = NULL;
+  for (size_t i = 1; i < idcount; i++) {
+    tmpid = sk_OCSP_SINGLERESP_value(sresp, i)->certId;
+    // Check to see if IDs match
+    if (OCSP_id_issuer_cmp(cid, tmpid) != 0) {
+      // If algorithm mismatch, let caller deal with it instead
+      if (OBJ_cmp(tmpid->hashAlgorithm->algorithm, cid->hashAlgorithm->algorithm) != 0) {
+        return 2;
+      }
+      return 0;
+    }
+  }
+
+  // All IDs match: only need to check one ID
+  *ret = cid;
+  return 1;
+}
+
+static int ocsp_match_issuerid(X509 *cert, OCSP_CERTID *cid,
+                               STACK_OF(OCSP_SINGLERESP) *sresp) {
+  if (cert == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  // If only one ID to match then do it.
+  if (cid) {
+    const EVP_MD *dgst;
+    X509_NAME *iname;
+    unsigned char md[EVP_MAX_MD_SIZE];
+    // Set up message digest for comparison.
+    dgst = EVP_get_digestbyobj(cid->hashAlgorithm->algorithm);
+    if (dgst == NULL) {
+      OPENSSL_PUT_ERROR(OCSP, OCSP_R_UNKNOWN_MESSAGE_DIGEST);
+      return -1;
+    }
+    size_t mdlen = EVP_MD_size(dgst);
+    iname = X509_get_subject_name(cert);
+    if (!X509_NAME_digest(iname, dgst, md, NULL)) {
+      return -1;
+    }
+
+    // Compare message digest with |OCSP_CERTID|
+    if ((cid->issuerNameHash->length != (int)mdlen) || (cid->issuerKeyHash->length != (int)mdlen)) {
+      return 0;
+    }
+    if (memcmp(md, cid->issuerNameHash->data, mdlen) != 0) {
+      return 0;
+    }
+    if (0 <= X509_pubkey_digest(cert, dgst, md, NULL)) {
+      if (memcmp(md, cid->issuerKeyHash->data, mdlen) != 0) {
+        return 0;
+      }
+    }
+    return 1;
+
+  }
+  // We have to match the whole stack recursively, if |cid| is NULL. This means
+  // that the issuer certificate's hash algorithm did not match with the rest of
+  // the |certId|s in the |OCSP_SINGLERESP| stack. (Issuer certificate is taken
+  // from the root of the |OCSP_SINGLERESP| stack).
+  // We'll have to find a match from the signer or responder CA certificate (both
+  // are certificates in the certificate chain) in the |OCSP_SINGLERESP| stack
+  // instead.
+  else {
+    int ret;
+    OCSP_CERTID *tmpid;
+    for (size_t i = 0; i < sk_OCSP_SINGLERESP_num(sresp); i++) {
+      tmpid = sk_OCSP_SINGLERESP_value(sresp, i)->certId;
+      ret = ocsp_match_issuerid(cert, tmpid, NULL);
```

Comment:
Is `tmpid` guaranteed to be not `NULL`? If not, a not NULL check is needed on `sresp`.